### ISSUE TITLE
DLND-Week2(2)

### ISSUE THREAD

**Issue by chenghong-lin-nu** [Hclmaster/blog#4]
## Sirag的数据准备
- Standard data format is csv.
- Three key pre-processing steps for every data set we will cover:
- Cleaning, Transformation, Reduction

## 模型评估和验证
- 数据分为训练、验证和测试集
- 在此部分，我们将学习为何需要这么做，以及如何验证模型的效果并为其评分。

## Regression and Classification
- Regression是一个预测值的模型。
- 分类问题主要用来确定状态。
![image](https://user-images.githubusercontent.com/16880879/35714760-94cd37ac-0809-11e8-85d5-fa89a9ee09b1.png)
- 左边的Model generalize better
- 右边的Model太过于完美拟合数据，称为**过拟合（Overfitting）**
![image](https://user-images.githubusercontent.com/16880879/35714897-536c5aa8-080a-11e8-908d-092363a29802.png)
- 那么怎样找到一个generalize好的model呢？
- 我们把数据分为Train Set和Test Set。
- 可以看到虽然右边的模型适合Train Set更好，但是从Test Set的误差来看，还是左边的模型更好一点。
![image](https://user-images.githubusercontent.com/16880879/35714972-bc9d1a80-080a-11e8-891d-c120f49f40e2.png)
- 当然我们也可以将其用于Classification问题中。
- 异常值被视为噪音。
- 在sklearn中这样做很简单，使用函数train-test-split即可。
- **Golden Rule: Never use your testing data for training!**

## Evaluation Model
![image](https://user-images.githubusercontent.com/16880879/35717832-90f6cc98-081c-11e8-894c-ec87f272eb77.png)
![image](https://user-images.githubusercontent.com/16880879/35717967-32368986-081d-11e8-83c1-ff40cc72b254.png)
- **混淆矩阵(Confusion Matrix)** 只是储存这4个值的表格。

## Accuracy
- One of the ways to measure how good a model is.
- 其实就是正确的/总数

### 1. Mean Absolute Error
![image](https://user-images.githubusercontent.com/16880879/35718548-1287c494-0820-11e8-9606-79740ee1cb7d.png)

### 2. Mean Square Error(均方误差)
- 这里是点和线之间距离的平方。
![image](https://user-images.githubusercontent.com/16880879/35718657-8b2c690e-0820-11e8-8ee5-2cf09e8cd9af.png)

### 3. 常见的回归度量
- 称为"R2 Score"
- R2 Score通过将我们的模型与最简单的可能模型相比得出。
![image](https://user-images.githubusercontent.com/16880879/35719303-c6561e82-0823-11e8-82f5-c91921b12a98.png)
图片的上面是线性回归模型，下面是简单的模型。

### Oversimplifying the problem -> Underfitting(欠拟合)
- Doesn't do well in the training set.
- Error due to bias.
### Overcomplicating the problem -> Overfitting(过拟合)
- Does well in the training set, but it tends to memorize it instead of learning the characteristics of it.
- 在训练集表现好，但是在测试集就不行了。
- Error due to variance.
![image](https://user-images.githubusercontent.com/16880879/35719708-9b94ab9e-0825-11e8-952e-b34a0639c8b7.png)

## Model Complexity Graph(模型复杂性图表)
![image](https://user-images.githubusercontent.com/16880879/35720548-248a1b48-0829-11e8-8121-737cf7d6df73.png)
- 最左边的欠拟合，有high training and testing errors.
- 最右边的过拟合，training error很小，但是testing error很大。
- 所以从下图中可以看到fit最好的是二次
![image](https://user-images.githubusercontent.com/16880879/35733698-503b057e-0859-11e8-9392-61a933a27b39.png)
- 但是这里有一个问题，就是我们使用了testing data去训练。
- 所以这里可以将Training Data拆分成Training和Validation。
![image](https://user-images.githubusercontent.com/16880879/35734130-fff98890-085a-11e8-98d0-8d7785ed1585.png)
- UnderFit就像是没有准备好就去参加考试了；然后OverFit就像是你确实是准备了，但是你只是记住了那些细节却没有去理解，相当于是背诵，所以当考试一发生变化的时候你就不会了；而GoodFit就像是你在真正的学习。

## K-Fold Cross Validation(K折交叉验证)
- 待补（就是说有些可以用来train的数据我们不应该扔掉）。
